### ISSUE TITLE
Change apache listen port

### ISSUE THREAD

**Issue by egofont** [nextcloud/nextcloud-snap#134]
Hi.

Install Nextcloud on a raspberry pi3, with a NexcloudBox, I made the installation from zeros using a preinstalled version of ubuntu 16.04.

I have made everything work but for external access I need to do a port redirection on the router of my ISP provider. I can not leave the port 80 configured in apache since that port is used by the provider. When trying to change the apache configuration in the httpd.conf file located in /snap/nextcloud/173/conf/httpd.conf is in a read-only folder.

How can I do to change the listening port of apache?

Thank you.

**Comment by kyrofa** [comment]
First of all, can your router forward port 81 to port 80? That might be the easiest way to solve this problem.

The most recent release of snapd that just got SRUd back to xenial supports configuration, but I wanted to wait a little bit before I started utilizing that in case people were slow to update. Once I add support for it, you'll be able to change the port like this:

    snap set nextcloud apache_port=81

**Comment by egofont** [comment]
I can´t forward port 81 to port 80 on router. How i can update?

**Comment by Guillaume99** [comment]
I had more or less the same problem, and my solution was to use IPv6... so you will access directly your nextcloud box.

**Comment by dslachut** [comment]
**Additional use case for this feature: Reverse Proxy**

I'm currently running a regular Nextcloud install behind a Nginx reverse proxy. I'm migrating the host soon and I'd like to use the Snap distribution, but that won't work if the snap is locked to port 80.

**Comment by kyrofa** [comment]
Indeed, this is something we want to support. But we're currently blocked on LP: [#1636931](https://bugs.launchpad.net/snappy/+bug/1636931).

**Comment by roelbrand** [comment]
@kyrofa: it seems that as of 2017-01-05, the bug referenced above is closed.

**Comment by kyrofa** [comment]
Yeah now we're stuck on LP: [#1636934](https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1636934) for incorrectly saving config values even when the snap is removed (users won't be able to clear a config once they set it), and LP: [#1659924](https://bugs.launchpad.net/snappy/+bug/1659924) for not making hooks available for all users, in which case installation will either fail or config would be unusable for users who haven't run a dist-upgrade.

**Comment by Avnerus** [comment]
Hi! Maybe in the meantime you could just expose _httpd.conf_? 
I'm also waiting for this so I could install nextcloud behind nginx.

**Comment by Avnerus** [comment]
bump
Any other suggestion on installing behind nginx?

**Comment by kyrofa** [comment]
Note that both mentioned bugs are fixed in snapd trunk. Just waiting for a release.

**Comment by Avnerus** [comment]
@kyrofa It seems the new snapd version is released! Will get a nextcloud snap update soon?
Thank you.

**Comment by kyrofa** [comment]
@Avnerus indeed, snapd v2.24 contained what we were waiting for. This bug is on the queue.

**Comment by Avnerus** [comment]
@kyrofa , great! Thank you.

**Comment by kyrofa** [comment]
To anyone following this issue: please take a look at #266. There are links to snaps and instructions on how to test them. If you could please do so and comment on the pull request, I'd really appreciate it! I'd like to be sure that the fix solves the issue before releasing it.

**Comment by Avnerus** [comment]
@kyrofa Thank you! I am away right now but will be able to test it early next week.

**Comment by kyrofa** [comment]
Sorry everyone: we discovered LP: [#1690906](https://bugs.launchpad.net/snapd/+bug/1690906) in testing, which breaks upgrades for snaps installed in lxc. We need to hold off on this until that bug is fixed.

**Comment by kwauchope** [comment]
Not sure if this is helpful but a heads up in case it is (I came across this looking for another issue). The [commit](https://git.launchpad.net/snapd/commit/?id=60b356b1da0651f6904aa11b01e1f52c1a42d213) referenced [here](https://bugs.launchpad.net/snapd/+bug/1658140) sets the only valid characters in configure options to 'a-z, 0-9, -'. I notice the now reverted pull request for this uses a '.' character. The 'var validKey' regex still exists in the master of snap.

**Comment by a-m-s** [comment]
Any news on this? I want to ditch the somewhat-heavyweight lxd workaround and run nextcloud in a snap on the host, and the port conflict is the only thing stopping me.

**Comment by kyrofa** [comment]
Alright, we're gonna try this again. LP: [#1690906](https://bugs.launchpad.net/snapd/+bug/1690906) hasn't budged in status, but in testing it appears to have disappeared. We'll of course give it some heavy testing again. We should add some automated tests as well, but I put up #349 if folks want to test it out.

**Comment by kyrofa** [comment]
All tests (both in and out of lxc) look good. The PR has landed, which means this feature will soon be available in beta.

**Comment by vitaminj** [comment]
To avoid other people having to rummage, this is now implemented such that you can do:
```
sudo snap set nextcloud ports.http=8080
sudo systemctl restart snap.nextcloud.apache.service
```

**Comment by bor8** [comment]
HTTPS:

```
sudo snap set nextcloud ports.https=4444
sudo systemctl restart snap.nextcloud.apache.service
sudo apt update && sudo apt install nginx-extras
…
```
